### PR TITLE
Options.py typing

### PR DESCRIPTION
**Pull request by beauxq**
use `NumericOption` class inheriting from `numbers.Integral` instead of `int`
also `Range` can sometimes take text like:
"high": high end of range
"low": low end of range
~"true", "on": default if it exists, otherwise high end of range~
~"false", "off": zero if zero is the low end~

These are the typing changes I made to deal with issues in https://github.com/ArchipelagoMW/Archipelago/pull/356

### CODE REVIEW

**Comment by beauxq** [comment]
Just to copy some of the context from the other discussion here:

I ran into a bug that came from some of the `Option` classes inheriting from `int`
The issue comes from the fact that, in Python, `int` is immutable. So if we inherit from `int`, there is a hidden part of the object that is immutable.

```python3
class Option(int):
    value: int

    def __int__(self) -> int:
        return self.value


o = Option(3)

o.value = 5

for _ in range(o):
    print("hello")
```

Many people would expect that this code would print "hello" 5 times. But it does not.

I wouldn't be surprised if this has been causing some obscure bugs in Archipelago for a long time.

This is why I removed the inheritance from `int` and replaced it with the base class `NumericOption`

File: Options.py [hunk 1]
```diff
@@ -260,6 +407,16 @@ def from_text(cls, text: str) -> Range:
                 return cls(random.randint(cls.range_start, cls.range_end))
             else:
                 raise Exception(f"random text \"{text}\" did not resolve to a recognized pattern. Acceptable values are: random, random-high, random-middle, random-low, random-range-low-<min>-<max>, random-range-middle-<min>-<max>, random-range-high-<min>-<max>, or random-range-<min>-<max>.")
+        elif text in {"on", "true", "default", "yes", "high"}:
```
Comment (Berserker66) [inline Options.py:410]: I don't like adding all these names, I think just "low", "high", "default" and maybe "middle" (to reflect the random ranges) is plenty sufficient.

Comment (beauxq) [reply Options.py:410]: The motivation for this was for changing a setting from a toggle to a range.

Even if we don't end up changing the progression balancing from toggle to range, there may be something else in the future that makes that change.

If someone has a yaml file they have been using for a long time with `progression_balancing: on`, and that changes to a range, do you want to give them an error and force them to change that to something else, when we could just give them the default?

Comment (Berserker66) [reply Options.py:410]: Then the Prog Balancing subclass can have those additions, or a "transition" subclass, no reason to "pollute" every range.

**Pull request merged**